File: lab25_regras_associacao_apriori.R
# Title     : REGRAS DE ASSOCIAÇÃO
# Objective : Conhecer com a linguagem R pode potencializar a mineração de dados com foco na construção de ASSOCIAÇÕES usando o algoritmo APRIORI
# Created by: accol
# Created on: 09/06/2020

# Esta prática exige a instalação do pacote "arules"
# As regras serão carregadas a partir de um arquivo .txt no formato de cesta de compras, onde cada linha representa uma transação e os produtos são separados por vírgulas


library(arules)
library(arulesViz)

transacoes = read.transactions('D:/Users/Angelo/AULAS/Dados/transacoes.txt', format = 'basket', sep = ',')
transacoes
# Para uma melhor inspeção nas nossas transações podemos usaro o inspect
inspect(transacoes)
# Podemos criar uma imagem do arquivo para uma inspeção visual
image(transacoes)
# A partir deste momento vamos iniciar a mineração a partir da criação das regras. Observe que para restringir um pouco estou definindo que quero regras apenas com supporte e confiança mínimos de 50%. Sem essa condição muitas regras serão geradas => fique atento
regras = apriori(transacoes, parameter = list(supp=0.5, conf=0.5))
# Como podemos observar as regras geradas com mais detalhes => podemos usar o inspect
inspect(regras)
# Note que as primeiras regras não são relevantes => as regras interessantes surgem a partir da Regra de número 4
# Regras geradas o próximo passo é aplicá-las ao seu sistema de e-commerce, por exemplo
# Para melhorar nossa visualização vamos instalar o pacote "arulesViz"
plot(regras)
# Melhorando a visualização
plot(regras, method = 'graph', control = list(type = 'items'))

File: lab25_regras_associacao_apriori.py
# Title     : REGRAS DE ASSOCIAÇÃO
# Objective : Conhecer com a linguagem Python pode potencializar a mineração de dados com foco na construção de ASSOCIAÇÕES usando o algoritmo APRIORI
# Created by: accol
# Created on: 09/06/2020

# Precisaremos instalar um novo pacote => pip install apyori
# Esta prática exige a instalação do pacote "arules"
# As regras serão carregadas a partir de um arquivo .txt no formato de cesta de compras, onde cada linha representa uma transação e os produtos são separados por vírgulas


import pandas as pd 
from apyori import apriori


dados = pd.read_csv('D:/Users/Angelo/AULAS/Dados/transacoes.txt')
print('\nPodemos agora visualizar os dados\n', dados)
# Vamos agora eliminar o header
dados = pd.read_csv('D:/Users/Angelo/AULAS/Dados/transacoes.txt', header=None)
print('\nPodemos agora visualizar os dados sem header\n', dados)
# Precisamos agora fazer alterações na variável para ajustar os cálculos
transacoes = []
for i in range(0,6):
    transacoes.append([str(dados.values[i,j]) for j in range(0,3)])
print('\nVamos observar a variável transações\n', transacoes)

# Vamos agora gerar as regras -> vamos ajustar suporte e confiança para 0.5
regras = apriori(transacoes, min_support = 0.5, min_confidence = 0.5)
resultados = list(regras)
print('\nAs regras geradas são:\n', resultados)
resultados2 = [list(x) for x in resultados]
print('\nAs regras geradas melhoradas são:\n', resultados2)
resultados3 = []
for j in range(0,7):
    resultados3.append([list(x) for x in resultados2[j][2]])
print('\nAs regras geradas melhoradas são:\n', resultados3)
